File: tags/CARDAMOM_V1_1_CE/Repository/simrepository/c++/src/SimProcessCallback_impl.cpp
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


#include "simrepository/SimProcessCallback_impl.hpp"
#include "SystemMngt/idllib/CdmwPlatformMngtProcess.stub.hpp"
#include <sstream>
#include <iomanip>

namespace
{
    const char* PROCESS_PREFIX="PROC_";
};

namespace Cdmw
{
namespace Tools
{


SimProcessCallback_impl::SimProcessCallback_impl(
    PortableServer::POA_ptr                     parent,
    CosNaming::NamingContext_ptr                root,
    CdmwPlatformMngt::ProcessMessageBroker_ptr message_broker,
    CdmwPlatformMngtEntity::EntityObserver_ptr entity_observer,
    CdmwPlatformMngtService::ServiceBroker_ptr service_broker,
    CdmwSimulatedRepository::MessageLogger_ptr message_logger,
    const std::string &                         application_name)
    throw (CORBA::SystemException)
    : m_poa(PortableServer::POA::_duplicate(parent)),
      m_root(root),
      m_message_broker(CdmwPlatformMngt::ProcessMessageBroker::_duplicate(message_broker)),
      m_entity_observer(CdmwPlatformMngtEntity::EntityObserver::_duplicate(entity_observer)),
      m_service_broker(CdmwPlatformMngtService::ServiceBroker::_duplicate(service_broker)),
      m_logger(CdmwSimulatedRepository::MessageLogger::_duplicate(message_logger)),
      m_application_name(application_name),
      m_process_number(0L)
{
}

SimProcessCallback_impl::~SimProcessCallback_impl() throw()
{
}
    
PortableServer::POA_ptr SimProcessCallback_impl::_default_POA()
{
    return PortableServer::POA::_duplicate(m_poa.in());
}


//
// IDL:thalesgroup.com/CdmwPlatformMngt/ProcessCallback/set_ready:1.0
//
CdmwPlatformMngt::ProcessCallback::ProcessAck*
SimProcessCallback_impl::set_ready(CdmwPlatformMngt::Process_ptr the_process)
    throw(CORBA::SystemException)
{
    CdmwPlatformMngt::ProcessCallback::ProcessAck ack;

    // The name of the application owning the process.
    ack.application_name= CORBA::string_dup(m_application_name.c_str());
    // The process name.
    {
        using namespace std;
        ostringstream ostrm;
        ostrm << PROCESS_PREFIX << setw(3) << setfill('0') << m_process_number;
        ++m_process_number;
        ack.process_name =  CORBA::string_dup(ostrm.str().c_str());        
    }
    // The observer used by the managed process to change the status of an entity.
    ack.entity_observer = m_entity_observer;
    // The broker used by the managed process to access services
    // that have been defined for it.
    ack.service_broker = m_service_broker;
    // The broker used by the managed process to notify its
    // applicative messages and fatal errors.
    ack.process_message_broker = m_message_broker;

    std::string body(m_application_name);
    body += "/";
    body += ack.process_name.in();
    m_logger->log("ProcessCallback::set_ready",body.c_str());
    
    // Register the_process in the naming context
    try {
        m_root.bind(body,the_process,true);
    } catch (...) {
        throw CORBA::INTERNAL();
    }

    return new CdmwPlatformMngt::ProcessCallback::ProcessAck(ack);
}

//
// IDL:thalesgroup.com/CdmwPlatformMngt/ProcessCallback/notify_error:1.0
//
void
SimProcessCallback_impl::notify_error(const char* issuer,
                                      const char* error_info)
    throw(CORBA::SystemException)
{
    std::string body("(");
    body += issuer;
    body += ",";
    body += error_info;
    body += ")";
    m_logger->log("ProcessCallback::notify_error",body.c_str());
}

} // namespace Tools
} // namespace Cdmw

File: tags/start/Repository/simrepository/java/src/cdmw/tools/SimProcessCallbackImpl.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES 
 * and SELEX-SI. 
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003. 
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under 
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your 
 * option) any later version. 
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public 
 * License for more details. 
 * 
 * You should have received a copy of the GNU Library General 
 * Public License along with CARDAMOM; see the file COPYING. If not, write to 
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.tools;

import cdmw.namingandrepository.NamingInterface;

import com.thalesgroup.CdmwPlatformMngt.ProcessCallbackPackage.ProcessAck;

/**
 * Provides an implementation of the ProcessCallback interface
 *
 * Features:
 * Thread safe
 *
 */
public class SimProcessCallbackImpl
    extends com.thalesgroup.CdmwPlatformMngt.ProcessCallbackPOA {

    private static final String PROCESS_PREFIX="PROC_";

    /**
     * Holds a reference to ...
     */
    private org.omg.PortableServer.POA poa;
    private NamingInterface root;
    private com.thalesgroup.CdmwPlatformMngt.ProcessMessageBroker messageBroker;
    private com.thalesgroup.CdmwPlatformMngtEntity.EntityObserver entityObserver;
    private com.thalesgroup.CdmwPlatformMngtService.ServiceBroker serviceBroker;
    private com.thalesgroup.CdmwSimulatedRepository.MessageLogger logger;
    private String applicationName;
    private int processNumber = 0;

    /**
     * Constructor
     * 
     */ 
    public SimProcessCallbackImpl(org.omg.PortableServer.POA parent,
        org.omg.CosNaming.NamingContext root,
        com.thalesgroup.CdmwPlatformMngt.ProcessMessageBroker messageBroker,
        com.thalesgroup.CdmwPlatformMngtEntity.EntityObserver entityObserver,
        com.thalesgroup.CdmwPlatformMngtService.ServiceBroker serviceBroker,
        com.thalesgroup.CdmwSimulatedRepository.MessageLogger logger,
        String applicationName) {
        
        this.poa = parent;
        this.root = new NamingInterface(root);
        this.messageBroker = messageBroker;
        this.entityObserver = entityObserver;
        this.serviceBroker = serviceBroker;
        this.logger = logger;
        this.applicationName = applicationName;
        
    }
    
    public org.omg.PortableServer.POA _default_POA() {
        return this.poa;
    }

    //
    // IDL:thalesgroup.com/CdmwPlatformMngt/ProcessCallback/set_ready:1.0
    //
    /**
     * Not thread safe!
     */
    public com.thalesgroup.CdmwPlatformMngt.ProcessCallbackPackage.ProcessAck
        set_ready(com.thalesgroup.CdmwPlatformMngt.Process theProcess) {
        
        ProcessAck ack = new ProcessAck();
        
        // The name of the application owning the process.
        ack.application_name = this.applicationName;
        // The process name.
        char[] procNumber = String.valueOf(this.processNumber).toCharArray();
        char[] procNb = { '0', '0', '0' };
        int i = 0;
        while(i < procNumber.length) {
            procNb[procNb.length - procNumber.length + i] = procNumber[i];
            i++;
        }
        
        processNumber++;
        ack.process_name = PROCESS_PREFIX + String.valueOf(procNb);
        // The observer used by the managed process to change the status of an entity.
        ack.entity_observer = this.entityObserver;
        // The broker used by the managed process to access services
        // that have been defined for it.
        ack.service_broker = this.serviceBroker;
        // The broker used by the managed process to notify its
        // applicative messages and fatal errors.
        ack.process_message_broker = this.messageBroker;

        String body = this.applicationName + "/" + ack.process_name;
        logger.log("ProcessCallback::set_ready", body);

        // Register theProcess in the naming context
        try {
            root.bind(body, theProcess, true);
        } catch (Exception e) {
            logger.log(e.toString(),"Unable to bind the process " 
                + ack.process_name + " with name " + body);
        }
        
        return ack;
        
    }

    //
    // IDL:thalesgroup.com/CdmwPlatformMngt/ProcessCallback/notify_error:1.0
    //
    public void notify_error(String issuer,String errorInfo) {
    
        String body = "(" + issuer + "," + errorInfo + ")";
        logger.log("ProcessCallback::notify_error", body);
    
    }

}

